### PR TITLE
Add icingadb support

### PR DESCRIPTION
**Pull request by sukhwinder33445**
requires https://github.com/Icinga/icingadb-web/pull/495
requires https://github.com/Icinga/ipl-web/pull/69

### CODE REVIEW

**Comment by HOSTED-POWER** [comment]
Is the support there already? We've been using graphite for long time and now with icingadb rc2 , we'd really need it supported :)

**Comment by HOSTED-POWER** [comment]
I tried this patch, but I still get this for new services (old ones are properly working!! ):

No such service

#0 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Graphite\Controllers\GraphController->serviceAction()
#1 /usr/share/php/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch()
#2 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch()
#3 /usr/share/php/Icinga/Application/Web.php(304): Zend_Controller_Front->dispatch()
#4 /usr/share/php/Icinga/Application/webrouter.php(107): Icinga\Application\Web->dispatch()
#5 /usr/share/icingaweb2/public/index.php(4): require_once(String)
#6 {main}

**Comment by sukhwinder33445** [comment]
Hi @HOSTED-POWER,

Icingadb support for this module is still in progress. More commits should be added soon.
Support will be available once this PR is merged.
Stay tuned :)

Thanks 
Sukhwinder

**Comment by HOSTED-POWER** [comment]
I see some updates, would that make new graphs work? 👍

**Comment by HOSTED-POWER** [comment]
Ok it did something for sure, now all graphs cannot be found anymore (before only new graphs were unfound)

 So I assume it uses the icingadb, but someting else might be wrong? Maybe some config missing or similar

**Comment by sukhwinder33445** [comment]
Hi @HOSTED-POWER,

This branch is still a WIP. There will be some more updates soon.
To test these changes, you need to do the following:

1. Set Icingadb as backend. (`Configuration` -> `Modules` -> `icingadb` -> `Backend tab`)

This should work.
 Please give feedback. If you get some errors or stack traces. Please add screenshots here.

**Comment by HOSTED-POWER** [comment]
Set Icingadb as backend. (Configuration -> Modules -> icingadb -> Backend tab)

-> You mean

 (Configuration -> Modules -> graphite -> Backend tab) ?

Since I cannot find the Icingadb switch under 

 (Configuration -> Modules -> icingadb -> Backend tab)

And well it does something, since the update no graphs are shown at all, before all old graphs worked, but new ones not.

I checked difference in URL:

OLD URL OF GRAPH (working):
https://ourmonitoringhost/graphite/graph/service?host.name=web1.xxxx.com&service.name=redis-default&service.check_command=redis&perfdata=hitrate&default_template=default-service&start=-3600&end=&width=1537&height=220&cachebuster=107762554145414&legend=1


AFTER ENABLING NEW MODULE WITH PATCHHES (not working):
https://ourmonitoringhost/graphite/graph/service?host.name=web1_xxxx_com&service.name=redis-default&service.check_command=redis&perfdata=hitrate&host=web1.xxxx.com&service=redis-default&default_template=default-service&start=-3600&end=&width=1537&height=220&cachebuster=107762445393425&legend=1

The _ with the hostname is strange?

**Comment by sukhwinder33445** [comment]
Please make sure that you have activated the [icingadb module](https://github.com/Icinga/icingadb-web). 
Then you can find it (Configuration -> Modules -> icingadb -> Backend tab).

**Comment by HOSTED-POWER** [comment]
I only have this backend for the moment, the monitoring backend is disabled completely (which also gave me a dependency missing for your graphite module)

backend1

backend2

**Comment by HOSTED-POWER** [comment]
I see this when I replace the URL:

Uncaught TypeError: Argument 2 passed to Icinga\Module\Graphite\Graphing\Template::getCharts() must be an instance of Icinga\Module\Monitoring\Object\MonitoredObject, instance of Icinga\Module\Icingadb\Model\Service given, called in /usr/share/icingaweb2/modules/graphite/library/Graphite/Web/Widget/Graphs.php on line 262 and defined in /usr/share/icingaweb2/modules/graphite/library/Graphite/Graphing/Template.php:73
Stack trace:
#0 /usr/share/icingaweb2/modules/graphite/library/Graphite/Web/Widget/Graphs.php(262): Icinga\Module\Graphite\Graphing\Template->getCharts()
#1 /usr/share/icingaweb2/modules/graphite/library/Graphite/Web/Widget/Graphs.php(399): Icinga\Module\Graphite\Web\Widget\Graphs->getGraphsList()
#2 /usr/share/php/Icinga/Web/Widget/AbstractWidget.php(114): Icinga\Module\Graphite\Web\Widget\Graphs->render()
#3 /usr/share/icinga-php/ipl/vendor/ipl/html/src/Text.php(62): Icinga\Web\Widget\AbstractWidget->__toString()
#4 /usr/share/icinga-php/ipl/vendor/ipl/html/src/Text.php(28): ipl\Html\Text->setCon

#0 [internal function]: Icinga\Application\Web->Icinga\Application\{closure}()
#1 {main}

However after refreshing the page , I just see empty images again:


https://ourmonitoringhost/graphite/graph/service?host.name=mgmt_xxxx_cloud&service.name=_var_lib_elasticsearch&service.check_command=disk&disk=_var_lib_elasticsearch&host=mgmt.xxxx.cloud&service=%2Fvar%2Flib%2Felasticsearch&template=disk&start=-3600&end=&width=350&height=200&cachebuster=107762850625081&legend=1&r=1644330608808

It really looks like the URL as invalid and not found

Another invalid image:

https://ourmonitoringhost/graphite/graph/service?host.name=shops23_xxx_com&service.name=apt&service.check_command=apt&host=shops23.xxx.com&service=apt&template=apt&start=-3600&end=&width=350&height=200&cachebuster=107762850635106&legend=1&r=1644330769081

Giving:

No such service

#0 /usr/share/icingaweb2/modules/graphite/application/controllers/GraphController.php(93): Icinga\Module\Graphite\Controllers\GraphController->icingadbServiceAction()
#1 /usr/share/icingaweb2/library/vendor/Zend/Controller/Action.php(507): Icinga\Module\Graphite\Controllers\GraphController->serviceAction()
#2 /usr/share/php/Icinga/Web/Controller/Dispatcher.php(76): Zend_Controller_Action->dispatch()
#3 /usr/share/icingaweb2/library/vendor/Zend/Controller/Front.php(937): Icinga\Web\Controller\Dispatcher->dispatch()
#4 /usr/share/php/Icinga/Application/Web.php(304): Zend_Controller_Front->dispatch()
#5 /usr/share/php/Icinga/Application/webrouter.php(107): Icinga\Application\Web->dispatch()
#6 /usr/share/icingaweb2/public/index.php(4): require_once(String)
#7 {main}

However I fixed it by manually modifying the URL to:


https://ourmonitoringhost/graphite/graph/service?host.name=shops23.xxx.com&service.name=apt&service.check_command=apt&host=shops23.xxx.com&service=apt&template=apt&start=-3600&end=&width=350&height=200&cachebuster=107762850635106&legend=1&r=1644330769081

Basically  I changed the host.name to . seperator instead of _ (so shops23_xxx_com -> shops23.xxx.com)

I did this now with all non working URL's (as a test in browser) and all the sudden I can see the graph, so this must be the cause!! 

Could you look into this? And probably need to replace the _ to . somehow in your code :)

File: application/controllers/IcingadblistController.php [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+        $this->addContent((new IcingadbGraph($services)));
+
+        $this->setAutorefreshInterval(10);
+    }
+
+    public function completeAction()
+    {
+        $suggestions = new ObjectSuggestions();
+        $suggestions->setModel(Service::class);
```
Comment (nilmerg) [inline application/controllers/IcingadblistController.php:153]: This way it's only for services. For hosts, invalid suggestions may be the result.

File: application/controllers/IcingadblistController.php [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        $compact = $this->view->compact;
+
+        $db = $this->getDb();
+
+        $hosts = Host::on($db)->with(['state']);
+        // I think, we don't need this, as we fetch the image data from graphite and not from redis
+        $hosts->setResultSetClass(VolatileStateResults::class);
+
+        $hosts->getSelectBase()
+            ->where(['performance_data != ?' => '']);
```
Comment (nilmerg) [inline application/controllers/IcingadblistController.php:43]: Another case for `$hosts->filter()` ;)

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Graphite/Graphing/MetricsQuery.php [hunk 3]
```diff
@@ -118,22 +122,35 @@ public function fetchRow()
     public function fetchColumn()
     {
         $filter = [];
+
```
[the remark this answers is not in the document]
Comment (sukhwinder33445) [reply library/Graphite/Graphing/MetricsQuery.php:125]: This was not my intention and was a mistake, sorry. I added some vardumps for debugging and then removed them, but not the new lines created by vardumps.

**Comment by HOSTED-POWER** [comment]
I did more tests and it's consistent, it's just the host.name which is wrong, it uses _ instead of . for the fqdn names :)

Is this easy to fix? It would be a good start for us and would give us all graphs back!

File: library/Graphite/Web/Controller/MonitoringAwareController.php [hunk 4]
```diff
@@ -20,4 +24,14 @@ protected function applyMonitoringRestriction(DataView $dataView)
...
     }
+
+    protected function moduleInit()
+    {
+        if (Module::exists('icingadb') && IcingadbSupport::useIcingaDbAsBackend()) {
+            return;
+        }
+
+        $this->backend = Backend::createBackend($this->_getParam('backend'));
+        $this->view->url = Url::fromRequest();
```
[the remark this answers is not in the document]
Comment (sukhwinder33445) [reply library/Graphite/Web/Controller/MonitoringAwareController.php:35]: Because I use the same ListController for the URL, an error is thrown if no IdoBackend resource is defined (No Backend set). Therefore, I override the method so that the controller does not need the Ido backend resource if icingadb is set.

File: application/controllers/IcingadblistController.php [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+    public function hostsAction()
+    {
+        $this->setTitle(t('Hosts'));
+        $compact = $this->view->compact;
+
+        $db = $this->getDb();
+
+        $hosts = Host::on($db)->with(['state']);
+        // I think, we don't need this, as we fetch the image data from graphite and not from redis
```
[the remark this answers is not in the document]
Comment (sukhwinder33445) [reply application/controllers/IcingadblistController.php:39]: We need performance data to select only objects with performance data. This performance data has no influence on the graph. I cannot find any use of performance data in the Graphs class.

File: library/Graphite/Web/Controller/MonitoringAwareController.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Graphite/Web/Controller/MonitoringAwareController.php:35]: Aaaah.. and why don't you call the parent method then?

File: application/controllers/IcingadblistController.php (hunk 1 above)
Comment (nilmerg) [reply application/controllers/IcingadblistController.php:153]: It may work, but it's still wrong.

>  Should I separate host and service actions? (Create a new class for each with a complete action for each).

Sounds like a good idea! You could also make these the default routes without the need to blacklist anything. The menu entries can be dynamically provided after all.

File: library/Graphite/Graphing/MetricsQuery.php [hunk 6]
```diff
@@ -118,22 +122,35 @@ public function fetchRow()
...
+                }
+                $result = $this->resolveMacro($workaroundMacro, $this->object);
+            } else {
+                $workaroundMacro = str_replace('.', '_', $macro);
+                if ($workaroundMacro === 'service_name') {
+                    $workaroundMacro = 'service_description';
+                }
+
+                $result = Macro::resolveMacro($workaroundMacro, $this->object);
+            }
```
[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Graphite/Graphing/MetricsQuery.php:153]: The possible macros. I can't review this part of the code without knowing which macros are possible/supported.

[the remark this answers is not in the document]
Comment (sukhwinder33445) [reply library/Graphite/Graphing/MetricsQuery.php:153]: I checked it again, this resolves the macros  `host|service.name` and `host|service.checkcommand`.

**Comment by sukhwinder33445** [comment]
> However I fixed it by manually modifying the URL to:

Is your hostname in this case defined with `.` in between?

**Comment by HOSTED-POWER** [comment]
> Hi @HOSTED-POWER
> 
> > However I fixed it by manually modifying the URL to:
> 
> What is the correct name of your host? Shops23.xxx.com" OR "Shops23_xxx_com"? What name do you see in the detailed view of the host? Can you add a screenshot if possible?

All our hosts use the FQDN of the machine, adding a sample here

example-host

Keeping fingers crossed for a fix hahaha :D

**Comment by HOSTED-POWER** [comment]
> What is the correct name of your host? Shops23.xxx.com" OR "Shops23_xxx_com"?

So it would be Shops23.xxx.com, any of the fixes you just committed to include this as well? Strange we end up with _ while the . was used :)

**Comment by sukhwinder33445** [comment]
Hello @HOSTED-POWER ,

The url parameters should work correctly now. Can you please test it again and give me feedback?

**Comment by sukhwinder33445** [comment]
It works on my system. 
Does it still not work on your system?  If not, what kind of error is occurring now? Still the same wrong INVALID url?

**Comment by HOSTED-POWER** [comment]
> Hello @HOSTED-POWER ,
> 
> The url parameters should work correctly now. Can you please test it again and give me feedback?

I just installed it now as a test and I can confirm all is working, wow, thanks a lot, really happy with this!!

**Comment by sukhwinder33445** [comment]
Thank you very much for your feedback. 
I am testing the changes now. Please give feedback if you find further errors or problems.

File: library/Graphite/Web/Controller/MonitoringAwareController.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Graphite/Web/Controller/MonitoringAwareController.php]: You've removed it entirely. It sets up the monitoring backend for no reason now. This happens in case of the redirects and for every single graph. Please add it again.

File: library/Graphite/Web/Widget/Graphs/Icingadb/IcingadbGraph.php [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        $this->objects = $objects;
+        $this->hostBaseUrl = $hostBaseUrl;
+        $this->serviceBaseUrl = $serviceBaseUrl;
+    }
+
+    protected function assemble()
+    {
+        if ($this->objects->count() === 0) {
```
Comment (nilmerg) [inline library/Graphite/Web/Widget/Graphs/Icingadb/IcingadbGraph.php:56]: If you'd require a `ipl\Orm\ResultSet` to be passed as `$objects`, you'd be able to call `->hasResult()` here which is more efficient. You'll get the result set by calling `->execute()` on the query. (In the controller!)

File: library/Graphite/ProvidedHook/Icingadb/ServiceDetailExtension.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+            $this->handleTimeRangePickerRequest();
+
+            $header = Html::tag('h2', [], 'Graphs');
+            $timepicker = HtmlString::create($this->renderTimeRangePicker(Icinga::app()->getViewRenderer()->view));
+            $graphs = HtmlString::create($graphs);
+
+            return HtmlString::create($header . $timepicker . $graphs);
+        }
+
+        return HtmlString::create(null);
```
Comment (nilmerg) [inline library/Graphite/ProvidedHook/Icingadb/ServiceDetailExtension.php:44]: Don't pass `null` here, but the empty string. The documentation says it accepts only strings. Once we enforce this by type hinting, this will break.

File: library/Graphite/ProvidedHook/Icingadb/HostDetailExtension.php [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+            $this->handleTimeRangePickerRequest();
+
+            $header = Html::tag('h2', [], 'Graphs');
+            $timepicker = HtmlString::create($this->renderTimeRangePicker(Icinga::app()->getViewRenderer()->view));
+            $graphs = HtmlString::create($graphs);
+
+            return HtmlString::create($header . $timepicker . $graphs);
+        }
+
+        return HtmlString::create(null);
```
Comment (nilmerg) [inline library/Graphite/ProvidedHook/Icingadb/HostDetailExtension.php:43]: Same as above. The empty string please.

File: library/Graphite/Graphing/MetricsQuery.php [hunk 10]
```diff
@@ -128,13 +132,25 @@ public function fetchColumn()
...
+            if ($this->object instanceof Model) {
+                // icingadb macros
+                $workaroundMacro = $macro;
+                if ($workaroundMacro === 'service.check_command') {
+                    $workaroundMacro = 'service.checkcommand';
+                }
+                if ($workaroundMacro === 'host.check_command') {
+                    $workaroundMacro = 'host.checkcommand';
+                }
+                $result = $this->resolveMacro($workaroundMacro, $this->object);
```
Comment (nilmerg) [inline library/Graphite/Graphing/MetricsQuery.php:144]: I fear that there may be templates out there, which use even more macros that only the monitoring module understands. Have you checked all the default templates for any macros that need a migration? Please also think of asking our colleagues in the #tech channel for their experiences regarding macros in templates.

File: configuration.php [hunk 11]
```diff
@@ -3,9 +3,19 @@
...
 /** @var \Icinga\Application\Modules\MenuItemContainer $section */
+
+use Icinga\Authentication\Auth;
+use Icinga\Module\Graphite\ProvidedHook\Icingadb\IcingadbSupport;
+
 $section = $this->menuSection(N_('Graphite'), ['icon' => 'chart-area']);
-$section->add(N_('Hosts'), ['url' => 'graphite/list/hosts?graphs_limit=2']);
-$section->add(N_('Services'), ['url' => 'graphite/list/services?graphs_limit=2']);
+
+if ($this::exists('icingadb') && Auth::getInstance()->getUser() !== null && IcingadbSupport::useIcingaDbAsBackend()) {
```
Comment (nilmerg) [inline configuration.php:12]: I see why you're checking for a user, but I think it's a bad idea to require another check for callers of `useIcingaDbAsBackend()`. The latter should automatically avoid accessing the preferences if there's no logged in user. Please change this in Icinga DB Web instead.

File: configuration.php [hunk 12]
```diff
@@ -3,9 +3,19 @@
...
+        $section->add(N_('Hosts'), ['url' => 'graphite/hosts?graphs_limit=2']);
+        $section->add(N_('Services'), ['url' => 'graphite/services?graphs_limit=2']);
```
Comment (nilmerg) [inline configuration.php:14]: The indentation looks somewhat off to me here :wink:

File: application/controllers/ServicesController.php [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        // shift graph params to avoid exception
+        foreach (['graphs_limit', 'graph_range', 'graph_start', 'graph_end'] as $param) {
+            $this->params->shift($param);
+        }
+
+        $this->setTitle(t('Services'));
+
+        $db = $this->getDb();
+
+        $services = Service::on($db)->with(['state']);
```
Comment (nilmerg) [inline application/controllers/ServicesController.php:33]: Are you calling `with` just because of the filter? If so, and you don't need access to the state afterwards, don't call it. You're needlessly fetching more data otherwise.

File: application/controllers/ListController.php [hunk 14]
```diff
@@ -29,6 +31,14 @@ public function init()
 
     public function hostsAction()
     {
+        if (Module::exists('icingadb') && IcingadbSupport::useIcingaDbAsBackend()) {
+            if ($hostName = $this->params->shift('host')) {
+                $this->params->set('name', $hostName);
```
Comment (nilmerg) [inline application/controllers/ListController.php:36]: Please use `host.name` here. Just `name` may work fine, but that's more of a convenience for lazy users typing this into the search bar :wink:

File: application/controllers/ListController.php [hunk 15]
```diff
@@ -67,6 +77,18 @@ public function hostsAction()
 
     public function servicesAction()
     {
+        if (Module::exists('icingadb') && IcingadbSupport::useIcingaDbAsBackend()) {
+            if ($hostName = $this->params->shift('host')) {
+                $this->params->set('host.name', $hostName);
+            }
+
+            if ($serviceName = $this->params->shift('service')) {
+                $this->params->set('name', $serviceName);
```
Comment (nilmerg) [inline application/controllers/ListController.php:86]: Also here, `service.name` please.

File: application/controllers/GraphController.php [hunk 16]
```diff
@@ -74,6 +89,10 @@ public function hostAction()
 
     public function serviceAction()
     {
+        if (Module::exists('icingadb') && IcingadbSupport::useIcingaDbAsBackend()) {
+            $this->icingadbService();
+            return;
+        }
```
Comment (nilmerg) [inline application/controllers/GraphController.php:95]: An empty line after the closing brace please.

File: application/controllers/ServicesController.php (hunk 13 above)
Comment (nilmerg) [reply application/controllers/ServicesController.php:33]: But... what if there's a macro like `service.state.timeout`. That would result in a query per object then. (25 additional queries per page) Not good. Maybe just forget my comment above, and even add `host` to this list, for macros like `host.name` :)

**Review by nilmerg — CHANGES_REQUESTED** [review]
* Don't forget my comment [here](https://github.com/Icinga/icingaweb2-module-graphite/pull/259#discussion_r807949814)
* Clearing the filter in the search editor redirects to icingadb/hosts or icingadb/services
* Clicking on any "Load all n graphs" results in an error

File: library/Graphite/Web/Widget/Graphs/Icingadb/IcingadbService.php [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+
+    public function createHostLink($baseUrl)
+    {
+        return $baseUrl->with(['name' => $this->object->host->name]);
+    }
+
+    public function createServiceLink($baseUrl)
+    {
+        return $baseUrl->with(['name' => $this->object->name, 'host.name' => $this->object->host->name]);
+    }
```
[the remark this answers is not in the document]
Comment (sukhwinder33445) [reply library/Graphite/Web/Widget/Graphs/Icingadb/IcingadbService.php:51]: `$baseUrl `contains the TimeRange parameters here when selected so that the same TimeRange can be selected when opening the object detail view.

**Comment by sukhwinder33445** [comment]
For the new changes to work, you must checkout the following repository branches:
icingadb-web : `introduce-sort-controls-trait`
ipl-web : `introduce-search-controls-trait`

File: library/Graphite/Web/Controller/IcingadbGraphiteController.php [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+{
+    use Auth;
+    use Database;
+    use SearchControls;
+
+    /** @var Filter\Rule Filter from query string parameters */
+    private $filter;
+
+    /** @var string|null */
+    private $format;
```
Comment (nilmerg) [inline library/Graphite/Web/Controller/IcingadbGraphiteController.php:29]: You shouldn't do this. It's not this module's duty to handle this parameter.

File: library/Graphite/Web/Controller/IcingadbGraphiteController.php [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+
+        return $limitControl;
+    }
+
+    public function filter(Query $query, Filter\Rule $filter = null): self
+    {
+        //TODO don't need format?
+        if ($this->format !== 'sql' || $this->hasPermission('config/authentication/roles/show')) {
+            $this->applyRestrictions($query);
+        }
```
Comment (nilmerg) [inline library/Graphite/Web/Controller/IcingadbGraphiteController.php:66]: Yes. Just apply the restrictions without any condition.

File: library/Graphite/Web/Controller/IcingadbGraphiteController.php [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+
+        if ($query instanceof UnionQuery) {
+            foreach ($query->getUnions() as $query) {
+                $query->filter($filter ?: $this->getFilter());
+            }
+        } else {
```
Comment (nilmerg) [inline library/Graphite/Web/Controller/IcingadbGraphiteController.php:72]: You don't feed this with any union query. Remove this.

Comment (nilmerg) [reply library/Graphite/Web/Controller/IcingadbGraphiteController.php:72]: And you don't feed this ever without a filter... I guess it's sufficient to use this in both controllers and to remove this method:

```
$this->applyRestrictions($query);
$query->filter($filter);
```

File: library/Graphite/Web/Controller/IcingadbGraphiteController.php [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+
+        return $sortControl->apply($query);
+    }
+
+    public function preDispatch()
+    {
+        parent::preDispatch();
+
+        $this->format = $this->params->shift('format');
+    }
```
Comment (nilmerg) [inline library/Graphite/Web/Controller/IcingadbGraphiteController.php:122]: Nah, again not your duty.

**Review by nilmerg — CHANGES_REQUESTED** [review]
* The routes `hosts` and `services` don't redirect to `list/hosts` and `list/services` respectively if icingadb is not the active backend.
* Don't forget the comment above which I've updated.

File: library/Graphite/Web/Widget/IcingadbGraphs.php [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+    /** Create a new Graph item
+     *
+     * @param Iterable $objects
+     *
+     * @param Url $hostBaseUrl Base url for hosts
+     *
+     * @param Url|null $serviceBaseUrl Base url for services
+     *
+     */
+    public function __construct($objects, $hostBaseUrl, $serviceBaseUrl = null)
```
[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Graphite/Web/Widget/IcingadbGraphs.php:45]: > not **just** *something* that's iterable.

Better?

[the remark this answers is not in the document]
Comment (sukhwinder33445) [reply library/Graphite/Web/Widget/IcingadbGraphs.php]: Got it ! Thanks

**Review by nilmerg — CHANGES_REQUESTED** [review]
The file `module.info` needs to require icingadb as well.

File: public/js/module.js [hunk 23]
```diff
@@ -109,4 +109,4 @@
 
     Icinga.Behaviors.Graphite = Graphite;
 
-})(Icinga);
```
[the remark this answers is not in the document]
Comment (nilmerg) [reply public/js/module.js:112]: yes

File: library/Graphite/Web/Controller/IcingadbGraphiteController.php [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+        $this->params->shift($sortControl->getSortParam());
+
+        return $sortControl->apply($query);
+    }
+
+    protected function moduleInit()
+    {
+        if (! Module::exists('icingadb') || ! IcingadbSupport::useIcingaDbAsBackend()) {
+            $this->useIcingadbAsBackend = false;
+        }
```
[the remark this answers is not in the document]
Comment (nilmerg) [reply library/Graphite/Web/Controller/IcingadbGraphiteController.php:107]: err, yes, as first method.

File: application/controllers/HostsController.php [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+            $params = base64_decode($this->params->get('legacyParams'));
+            $this->redirectNow(
+                Url::fromPath('graphite/list/hosts')
+                ->setParams(UrlParams::fromQueryString($params))
+            );
+        }
+
+        // shift graph params to avoid exception
+        $graphRange = $this->params->shift('graph_range');
+        $baseFilter = $graphRange ? QueryString::parse('graph_range=' . $graphRange) : $this->getFilter();
```
[the remark this answers is not in the document]
Comment (sukhwinder33445) [reply application/controllers/HostsController.php:39]: I passed "$this->filter()` so that setBaseFilter gets a Rule parameter, but now I see that it also accepts null. :)

**Comment by wornet-aer** [comment]
I've just checked out and tried your WIP code using icingaweb2 and icingadb.
Maybe this little feedback helps with testing.

In the service detail view the graphs are correctly displayed.

When I try to access the "Graphite -> Hosts" or Services from the menu, the following error is displayed:
```
Trait 'Icinga\Module\Icingadb\Common\SearchControls' not found
#0 [internal function]: Icinga\Application\Web->Icinga\Application\{closure}()
#1 {main}
```

File: application/controllers/ServicesController.php [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+
+    public function indexAction()
+    {
+        if (! $this->useIcingadbAsBackend) {
+            $params = urldecode($this->params->get('legacyParams'));
+            $this->redirectNow(Url::fromPath('graphite/list/services')->setQueryString($params));
+        }
+
+        $graphRange = $this->params->shift('graph_range');
+        $baseFilter = $graphRange ? QueryString::parse('graph_range=' . $graphRange) : $this->getFilter();
```
Comment (nilmerg) [inline application/controllers/ServicesController.php:30]: You know what to do here, right?

**Review by nilmerg — CHANGES_REQUESTED** [review]
Almost done :)

**Review by nilmerg — APPROVED** [review]
:+1: :tada:

File: application/controllers/HostsController.php [hunk 27]
```diff
@@ -0,0 +1,111 @@
...
+        }
+
+        // shift graph params to avoid exception
+        $graphRange = $this->params->shift('graph_range');
+        $baseFilter = $graphRange ? Filter::equal('graph_range', $graphRange) : null;
+        foreach ($this->graphParams as $param) {
+            $this->params->shift($param);
+        }
+
+        $this->setTitle(t('Hosts'));
```
Comment (nilmerg) [inline application/controllers/HostsController.php:37]: I forgot this. `setTitle` has changed. You'll need to call `addTitleTab` now. I've rebased https://github.com/Icinga/ipl-web/pull/69, otherwise `addTitleTab` is missing.

Comment (nilmerg) [reply application/controllers/HostsController.php:37]: The call to `setTitle` in method `searchEditorAction` does *not* need to change.

File: application/controllers/ServicesController.php [hunk 28]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        // shift graph params to avoid exception
+        $graphRange = $this->params->shift('graph_range');
+        $baseFilter = $graphRange ? Filter::equal('graph_range', $graphRange) : null;
+        foreach ($this->graphParams as $param) {
+            $this->params->shift($param);
+        }
+
+        $this->setTitle(t('Services'));
```
Comment (nilmerg) [inline application/controllers/ServicesController.php:35]: Also here, `addTitleTab` please.

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Review by nilmerg — APPROVED** [review]
So, but now!

**Comment by HOSTED-POWER** [comment]
> 

Is this still true? What is the best branch for this module? :)

**Comment by nilmerg** [comment]
This is in the latest release.

**Comment by HOSTED-POWER** [comment]
I meant for the icingadb web, sorry :)

**Comment by nilmerg** [comment]
Ah, master then.
